### PR TITLE
[move] Add interned string module move-symbol-pool

### PR DESCRIPTION
**Pull request by modocache**
The `Loc` type in Move uses a `&'static str` type to reference the name of the souce file its location refers to. Currently, the storage for that `str` filename is leaked: the reference exists for the duration of the program, but the storage is never freed.

To address this, add a string interning module that allocates global, static storage for strings, as well as a trivially copyable pointer to the string that can be used in the `Loc` type.

In #8742, this module's pointer type will be used in place of `&'static str' to represent `Loc::file` throughout the Move codebase.

### CODE REVIEW

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/lib.rs:26]: Thoughts on a `RwLock` v. `Mutex`?

I initially thought a `RwLock` would be better but the lock isn't actually used for reading so maybe the overhead is useless.

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/lib.rs:28]: Would you mind writing all of my documentation? lol 
This is glorious!

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/lib.rs:100]: maybe split these tests to a separate file?

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/lib.rs:69]: I kind of like this a lot more than using all the Rust traityness used by string_cache. All of those traits tend to hide what is going on

Having said that, it might be useful to do the optimization for `String` v. `&str` for `From` as you don't need to call `to_owned` for the `String` case

**Review by tnowacki — DISMISSED** [review]
Looks good to me! With the intention of beefing up the optimizations and perf and other things over time.

But, might want some others to look over it before landing. @wrwg and @metajack might both be interested.

**Comment by tnowacki** [comment]
Another thing, we might want to make sure we have consensus on the naming here. Maybe best to discuss this offline

**Comment by tnowacki** [comment]
Heads up from #8742: That PR really needs `From<String>` so maybe just add it to this one?

File: language/move-symbol-pool/src/lib.rs
Comment (wrwg) [inline language/move-symbol-pool/src/lib.rs:14]: I'm wondering why are we re-implementing our own (unsafe) intern solution. In the Move Prover we use the [internement](https://crates.io/crates/internment) crate successfully, so what's wrong with this one?

**Comment by wrwg** [comment]
> The `Loc` type in Move uses a `&'static str` type to reference the name of the souce file its location refers to. Currently, the storage for that `str` filename is leaked: the reference exists for the duration of the program, but the storage is never freed.
> 
> To address this, add a string interning module that allocates global, static storage for strings, as well as a trivially copyable pointer to the string that can be used in the `Loc` type.
> 
> In #8742, this module's pointer type will be used in place of `&'static str' to represent `Loc::file` throughout the Move codebase.

Another question: the Loc type in the Move compiler is outdated; newer versions of the this crate (which also the Move prover uses) have a FileId instead and a custom internalization of string -> FileId. So at least for this use case, isn't this home mode internalization obsolete, as we should anyway upgrade the Loc type?

**Comment by wrwg** [comment]
> > The `Loc` type in Move uses a `&'static str` type to reference the name of the souce file its location refers to. Currently, the storage for that `str` filename is leaked: the reference exists for the duration of the program, but the storage is never freed.
> > To address this, add a string interning module that allocates global, static storage for strings, as well as a trivially copyable pointer to the string that can be used in the `Loc` type.
> > In #8742, this module's pointer type will be used in place of `&'static str' to represent `Loc::file` throughout the Move codebase.
> 
> Another question: the Loc type in the Move compiler is outdated; newer versions of the this crate (which also the Move prover uses) have a FileId instead and a custom internalization of string -> FileId. So at least for this use case, isn't this home mode internalization obsolete, as we should anyway upgrade the Loc type?

FYI: this is the code in move-model where we adapt the legacy Loc used by move-lang to the more recent one: https://github.com/diem/diem/blob/main/language/move-model/src/model.rs#L730

**Comment by tnowacki** [comment]
> > > The `Loc` type in Move uses a `&'static str` type to reference the name of the souce file its location refers to. Currently, the storage for that `str` filename is leaked: the reference exists for the duration of the program, but the storage is never freed.
> > > To address this, add a string interning module that allocates global, static storage for strings, as well as a trivially copyable pointer to the string that can be used in the `Loc` type.
> > > In #8742, this module's pointer type will be used in place of `&'static str' to represent `Loc::file` throughout the Move codebase.
> > 
> > 
> > Another question: the Loc type in the Move compiler is outdated; newer versions of the this crate (which also the Move prover uses) have a FileId instead and a custom internalization of string -> FileId. So at least for this use case, isn't this home mode internalization obsolete, as we should anyway upgrade the Loc type?
> 
> FYI: this is the code in move-model where we adapt the legacy Loc used by move-lang to the more recent one: https://github.com/diem/diem/blob/main/language/move-model/src/model.rs#L730

Not quite sure what you mean by "legacy". There is no new one.
The `Loc` here has always been something handrolled and maintained by Move, so there has only ever been this version. 

The prover has something adapted to using FileId, but that is something from old versions of codespan reporting (new versions have dropped it).

The volatility around codespan reporting I think gives us all the more motivation to use our own interned file name here

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [reply language/move-symbol-pool/src/lib.rs:14]: The main crate we considered was servo's string_cache (https://docs.rs/string_cache/0.8.1/string_cache/). The problem with it is that it is ref counted. And all of our strings we parse are likely to live for the duration of the program, so the ref counting won't buy us much.

The internment crate is a nice general wrapper around `Box::leak`, but it is not really optimized for strings. 
We want to use this eventually for strings coming out of the parser (so any name/identifier a user gives). And about ~50% of those strings in the DF+Stdlib are less than 7 characters. There is a great optimization we can use from string_cache to stuff those characters into the actual symbol instead of allocating a string in the table (the bit stuffing code from string_cache: https://github.com/servo/string-cache/blob/master/src/atom.rs#L191)

In short, the existing crates don't quite match up with what we want to do long term, and the code here is rather simple to maintain and easy to optimize as we go forward

**Comment by wrwg** [comment]
> > > > The `Loc` type in Move uses a `&'static str` type to reference the name of the souce file its location refers to. Currently, the storage for that `str` filename is leaked: the reference exists for the duration of the program, but the storage is never freed.
> > > > To address this, add a string interning module that allocates global, static storage for strings, as well as a trivially copyable pointer to the string that can be used in the `Loc` type.
> > > > In #8742, this module's pointer type will be used in place of `&'static str' to represent `Loc::file` throughout the Move codebase.
> > > 
> > > 
> > > Another question: the Loc type in the Move compiler is outdated; newer versions of the this crate (which also the Move prover uses) have a FileId instead and a custom internalization of string -> FileId. So at least for this use case, isn't this home mode internalization obsolete, as we should anyway upgrade the Loc type?
> > 
> > 
> > FYI: this is the code in move-model where we adapt the legacy Loc used by move-lang to the more recent one: https://github.com/diem/diem/blob/main/language/move-model/src/model.rs#L730
> 
> Not quite sure what you mean by "legacy". There is no new one.
> The `Loc` here has always been something handrolled and maintained by Move, so there has only ever been this version.
> 
> The prover has something adapted to using FileId, but that is something from old versions of codespan reporting (new versions have dropped it).
> 
> The volatility around codespan reporting I think gives us all the more motivation to use our own interned file name here

You are right, Loc is our own. However, `FileId` is the way how files are handled [in the most recent codespan version I can find (0.11)](https://docs.rs/codespan/0.11.1/codespan/index.html). When you builld a Loc type based on codepan, it is natural to have FileId embedded in that type, instead of an internalized string. Eventually you need to construct that FileId anyway to get codespan to report.

I may stay corrected if there is a newer version of codespan. BTW, I only know about this one breaking change in codespan (namely when they introduced FileId), otherwise I'm happy with the stability.

**Comment by tnowacki** [comment]
> > > > > The `Loc` type in Move uses a `&'static str` type to reference the name of the souce file its location refers to. Currently, the storage for that `str` filename is leaked: the reference exists for the duration of the program, but the storage is never freed.
> > > > > To address this, add a string interning module that allocates global, static storage for strings, as well as a trivially copyable pointer to the string that can be used in the `Loc` type.
> > > > > In #8742, this module's pointer type will be used in place of `&'static str' to represent `Loc::file` throughout the Move codebase.
> > > > 
> > > > 
> > > > Another question: the Loc type in the Move compiler is outdated; newer versions of the this crate (which also the Move prover uses) have a FileId instead and a custom internalization of string -> FileId. So at least for this use case, isn't this home mode internalization obsolete, as we should anyway upgrade the Loc type?
> > > 
> > > 
> > > FYI: this is the code in move-model where we adapt the legacy Loc used by move-lang to the more recent one: https://github.com/diem/diem/blob/main/language/move-model/src/model.rs#L730
> > 
> > 
> > Not quite sure what you mean by "legacy". There is no new one.
> > The `Loc` here has always been something handrolled and maintained by Move, so there has only ever been this version.
> > The prover has something adapted to using FileId, but that is something from old versions of codespan reporting (new versions have dropped it).
> > The volatility around codespan reporting I think gives us all the more motivation to use our own interned file name here
> 
> You are right, Loc is our own. However, `FileId` is the way how files are handled [in the most recent codespan version I can find (0.11)](https://docs.rs/codespan/0.11.1/codespan/index.html). When you builld a Loc type based on codepan, it is natural to have FileId embedded in that type, instead of an internalized string. Eventually you need to construct that FileId anyway to get codespan to report.
> 
> I may stay corrected if there is a newer version of codespan. BTW, I only know about this one breaking change in codespan (namely when they introduced FileId), otherwise I'm happy with the stability.

Ah sorry yeah, I'm not referring to codespan itself really (we don't use it other than for the `Span` type), but for codespan reporting. In the recent version of that, `FileId` has been parameterized and abstracted away. So we could *might* be able to replace it with `Symbol`. It is now `struct Diagnostic<FileId>` https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/diagnostic/struct.Diagnostic.html
We would have to work out an implementation of this `Files` trait though: https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/files/trait.Files.html
But I don't think that will be too bad.

File: language/move-symbol-pool/src/lib.rs
Comment (modocache) [reply language/move-symbol-pool/src/lib.rs:28]: D'aw 😊 I was trying to learn about how docblocks work in Rust and had a lot of fun with it.

TBH I'm not sure all of this works right, I sort of copied the syntax I saw elsewhere for function/type references, and I don't think that syntax is validated by our tooling at the moment -- I'm definitely NOT suggesting it ought to be, but I assume these links are supposed to be used when HTML docs are generated, and who knows if that HTML will work with what I have here 😅

Comment (tnowacki) [reply language/move-symbol-pool/src/lib.rs:28]: I think there is a way to test this so you can see it, let me dig around

And I think any thing in quote blocks is validated? Or maybe we turned it off because it was too annoying? 
I know for a fact I had to turn it off here: https://github.com/diem/diem/blob/main/language/move-lang/src/shared/ast_debug.rs#L13

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-symbol-pool/src/lib.rs
Comment (modocache) [reply language/move-symbol-pool/src/lib.rs]: The latest implementation reads and writes, and so the `RwLock` is definitely an improvement -- thanks!

File: language/move-symbol-pool/src/lib.rs
Comment (modocache) [reply language/move-symbol-pool/src/lib.rs]: Yeah, I think the blocks are validated and tested -- at least, `cargo xtest -p move-symbol-pool` builds and runs them.

File: language/move-symbol-pool/src/pool.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/pool.rs:25]: nit never `if-let else` just `match` instead

File: language/move-symbol-pool/src/pool.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/pool.rs:23]: This seems... bad
Why not just use string_cache's table layout? I know you want to stay simple for the version version... but this is a bit too simple for my tastes

Comment (modocache) [reply language/move-symbol-pool/src/pool.rs:23]: Yeah, good call, I hadn't considered taking another look after making this latest change. Will do!

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-symbol-pool/src/pool.rs
Comment (modocache) [reply language/move-symbol-pool/src/pool.rs]: OK, the latest iteration is missing some tests that I'll add later today, but it uses string_cache's approach to storage 👍

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/lib.rs:30]: nit can you regroup this:
I'd expect 
- mods
- uses
- pub uses

File: language/move-symbol-pool/src/pool.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/pool.rs:86]: I think this needs to be reworded as, "different hash values always mean different strings"
Since Strings that are not equal _might_ have the same hash (hence the second check)

File: language/move-symbol-pool/src/symbol.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/symbol.rs:44]: wild... 

If this is true, should we just do `&'static str` Or are there lifetimes that aren't less than static?

File: language/move-symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/move-symbol-pool/src/lib.rs:4]: Should we point to the string_cache crate? would be good to give them a nod

**Review by tnowacki — COMMENTED** [review]
You will likely need another approval after you fix these comments, so I'll just approve after that. 

Looks good! thanks for all the iterations on this, I think we landed in a great initial spot!

File: language/move-symbol-pool/src/pool.rs
Comment (bmwill) [inline language/move-symbol-pool/src/pool.rs:60]: You can avoid the unsafe here with [into_boxed_slice](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.into_boxed_slice):
```suggestion
        Self(vec![None; NB_BUCKETS].into_boxed_slice())
```

File: language/move-symbol-pool/src/symbol.rs
Comment (bmwill) [inline language/move-symbol-pool/src/symbol.rs:41]: To improve ergonomics of using `Symbol` you could look into implementing `AsRef<str>`

File: language/move-symbol-pool/src/symbol.rs
Comment (bmwill) [reply language/move-symbol-pool/src/symbol.rs:44]: Might make more sense to bind the lifetime to `&self` that way the caller can't just arbitrarily choose the lifetime they want
```suggestion
    pub fn as_str(&self) -> &str {
```

File: language/move-symbol-pool/src/pool.rs
Comment (bmwill) [reply language/move-symbol-pool/src/pool.rs:60]: nvm i didn't realize you wanted a `Box<[Bucket; NB_BUCKETS]>`

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-symbol-pool/src/symbol.rs
Comment (modocache) [reply language/move-symbol-pool/src/symbol.rs]: I went with a `&self`-bound lifetime over a static one, but both type-check successfully. I don't know enough about Rust to have an opinion here. Since symbols are never removed from the pool, their underlying strings and references to them truly are static, and so AFAICT not necessarily bound to the lifetime of the `Symbol`.

File: language/move-symbol-pool/src/symbol.rs
Comment (modocache) [reply language/move-symbol-pool/src/symbol.rs:41]: Thanks! I added an implementation. I was bummed that it didn't allow me to remove any calls to `as_str` in #8742, though. I guess -- and since I'm still learning Rust hopefully you can teach me here -- that for that to have been possible, functions that take `&str` parameters could be rewritten to take `AsRef<str>` instead, where appropriate? I'll try to keep that in mind when writing my own functions from now on.

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-symbol-pool/src/symbol.rs
Comment (bmwill) [reply language/move-symbol-pool/src/symbol.rs:41]: Yep exactly that. There are tradeoffs* for making a function generic on `T: AsRef<str>` vs `&str` so i'm not saying you should prefer one over the other. Though it is beneficial to include the `AsRef` impl for the instances where you do have an API that takes in a `T: AsRef<str>`

\* Having lots of generic functions can increase compilation times and codesize since all instances are monomorphized. There are tricks to limiting that though, one such trick is having one API that takes in `T: AsRef<str>` where all its purpose is to call `as_ref()` to get the `&str` and then forward to another API that isn't generic and only takes `&str`s.

File: language/move-symbol-pool/src/symbol.rs
Comment (tnowacki) [reply language/move-symbol-pool/src/symbol.rs]: they are technically static as long as we don't drop anything.
Binding it to self is the best longterm decision if we ever do the symbol inlining thing

File: language/move-symbol-pool/src/symbol.rs
Comment (tnowacki) [reply language/move-symbol-pool/src/symbol.rs:41]: I had thought it would be bad to do a bunch of Rust traityness as it tends to hide behavior. But maybe it doesn't really matter here, and is best to be more ergonomic.

**Review by tnowacki — DISMISSED** [review]
Let's roll! Excited to start playing with this!

**Pull request closed without merging**
